### ISSUE TITLE
Request for a GPUQueue readBuffer API

### ISSUE THREAD

**Issue by benvanik** [gpuweb/gpuweb#1972]
During prototyping we may have hit a (practical) blocker in our use of WebGPU for complex compute workloads: the lack of persistently mapped buffers/exclusivity of access (buffers must be `unmapped` while any in-flight submission or queue operation may be using them) and the async-only GPUBuffer.mapAsync make writing efficient streaming programs (which need to perform pipelined readbacks and utilize buffer suballocation/ringbuffers that span multiple submissions) really difficult/special-casey. These are the same kind of situations that would call for features like PBOs in older graphics APIs, persistent mapping in newer ones, or things like cudaMemcpyAsync/clEnqueueReadBuffer/clEnqueueWriteBuffer in compute APIs.

Thankfully I think the missing API that would enable practical usage here may just be a pairing to GPUQueue writeBuffer that allows for a device timeline-synchronized readback into host memory:
```c
interface GPUQueue {
+   // (both could stand to have source/destination naming like copyBufferToBuffer does)
+   undefined readBuffer(
+       GPUBuffer buffer,
+       GPUSize64 bufferOffset,
+       [AllowShared] BufferSource data,
+       optional GPUSize64 dataOffset = 0,
+       optional GPUSize64 size);
    undefined writeBuffer(
        GPUBuffer buffer,
        GPUSize64 bufferOffset,
        [AllowShared] BufferSource data,
        optional GPUSize64 dataOffset = 0,
        optional GPUSize64 size);
}
```

Without this it's really tricky to do readback staging - something core to compute workflows - without either avoiding aliasing (usually a 10x+ memory increase - where we are talking on the order of 10MB->100-1000MB for things in the ML domain) or fully synchronizing with the device each frame. This makes it prohibitively expensive and difficult to target with portable code as when running in browser we'd need to both synchronize and use something like emscripten's asyncify during mapping (a non-starter for performance-sensitive workloads). Even if we were to natively write the JS to do this unless the entire program was carefully constructed (difficult when libraries/middleware is involved) we'd still have to synchronize as the mapAsync would fail if anything was in-flight and we couldn't submit more work against the buffers until the mapAsync had completed/we'd been called back, done our work on the buffer, and unmapped it.

The workaround I tried was to have a large number of readback staging buffers that could be in-flight at a time however management of the buffers, the double async pumps of waiting for submission completion + waiting for mapping, and the asyncify requirements pretty much killed that. The issues are compounded if there are multiple submissions in flight or multiple readback buffers required for a single combined submission (as it's not always possible to know when starting recording exactly how much needs to be read back and sometimes it's data-dependent). Simple workloads are ok _but_ simple workloads often run significantly better (lower latency, more predictable, lower API overhead, etc) on the CPU and avoid all this overhead so it's not a great way to show off WebGPU :)

If there were semaphores/fences on the queue we could use them to hide some of the latency involved in the readbacks by allowing independent work to proceed while copying things back, but given the single queue and sequential ordering we just have the hammer of stopping the world while we work and being beholden to the browser event loop in our critical path.

The unicorn alternative that is much more flexible would be even just a constrained allowance of persistent mapped memory such that a GPUCommandEncoder copyBufferToBuffer could copy from unmapped (device local) memory to mapped (host visible/local) memory with only the copy source/dest allowed to be mapped during submission. But I suspect there's reasons that's not allowed and GPUQueue was chosen as the place for writeBuffer to live (even if not someplace a compute-aware API would ever want it - host/device transfers are almost always interleaved with dispatches when pipelining). If there's a request tracker for persistent mapping, though, please let me know so I can +10000 that - it's about as useful to compute workloads as texture sampling is to graphics workloads :)

If there's something I'm missing in the spec and someone knows a way to perform pipelined readback of a GPUBuffer to host-accessible memory while there are pending/active in-flight commands operating on it I'd love to hear about it! I've read through a few times and can't spot anything that maps to what I know from other APIs, but I also kind of expect that with something new :)

/cc @dneto0 (this may be the only thing we can't easily work around - everything else looks doable)

**Comment by Kangz** [comment]
You're correct that the state of the buffer is for the complete buffer, so you'd use a ring of staging buffers instead of suballocating in one buffer. However I think there's a misunderstanding in the constraints around buffer mapping: buffers have to be in the "unmapped" state only during the submission because the GPU needs ownership, however right after the submit you can call `mapAsync` to enqueue an "GPU -> CPU ownership transfer" operation.

Code to do streaming of ML results could look like the following:

```js
var readbackBufferPool;

function frame() {
  var buffer = readbackBufferPool.takeOrCreateBuffer();
  var commands = encodeInference({resultGoesTo: buffer});
  device.queue.submit([commands]);

  // Start mapping now even as the commands are executing, so the
  // callback is fired as soon as the browser knows the GPU is done.
  buffer.mapAsync(...).then(() => {
    doSomethingWithResult(buffer.getMappedRange());
    buffer.unmap();
    readbackBufferPool.putBack(buffer);
  });

  requestAnimationFrame(frame);
}
```

An important design constraint of WebGPU is that it should integrate well with the Javascript event loop model, so it can't be blocking, and signals from the API have to come through the regular event loop. And right now, this means your application has to yield to the browser, on its own, or via ASYNCIFY. There has been discussions about making the application drive the browser event pump in Worker via stuff like a `requestMainLoop` but there has been no progress.

**Comment by benvanik** [comment]
Interesting! I think that highlights the core issue being that ownership transfer requirement: concurrent access from both devices is what is desired even if controlled by the queue via a readBuffer/writeBuffer vs. allowing true concurrent GPUBuffer access. Thanks for explaining as that was hard for me to detangle without it :)

The example you give works in very specific cases of exceedingly simple applications but is not something that scales well - especially not when libraries/middleware are involved: it's hard enough to get a simple portable codebase working in this arbitrary-async-in-the-middle-of-program-logic model, let alone having half a dozen that do everything precisely right and without non-trivial hoop jumping/performance issues. If we were talking about something like swap chain management that's fine as usually there's only one owner of such logic in a codebase, but memcpy is different. Things that are annoying/inconvenient in miniature often become impossible at scale :)

In general playing well with the event loop is fine for things that should need to play with the event loop (device enumeration, synchronizing with the device, etc). The argument here is that transferring memory device->host is not one of those cases that should need to as it is such a fundamental operation in compute workloads. As someone who used to primarily do graphics it's _also_ a fundamental operation in many graphics workloads but often far fewer are needed per (conceptual) frame - though I also ran into these same readback issues working on image/video encoding acceleration and more sophisticated rendering techniques (sparse virtual textures/geometry and custom queries - people around in the WebGL days may remember some of my issues ;). But in compute there really just aren't any workarounds: the 2 things you do are transfer memory to/from the device and execute operations on that memory and doing this kind of dance hundreds+ times is not viable.

Consider async/overlapped memcpy as a baseline requirement for an effective compute API. Given that writeBuffer already exists for host->device this is just WebGPU currently having a hole where the matching device->host equivalent should be (as if OpenCL only had clEnqueueWriteBuffer and no clEnqueueReadBuffer). After a bunch of trial I believe that WebGPU is really close to being a generally usable compute API (with some inefficiencies, but ones that can be evolved away) and this missing operation is the major blocker.

**Comment by Kangz** [comment]
> In general playing well with the event loop is fine for things that should need to play with the event loop (device enumeration, synchronizing with the device, etc). The argument here is that transferring memory device->host is not one of those cases that should need to as it is such a fundamental operation in compute workloads.

I'm not sure I understand, compute workloads have to wait on the GPU to read back the data anyway, so there needs to be some asynchrony. The other thing is that in a different issue you explained how you wanted to run the same workload many times and that some `GPUComputeBundle` would be helpful for it. Doesn't that mean that you know the size of your readback buffers in advance and can just use a pool mechanism as above?

Another reason why WebGPU doesn't have persistently mapped buffers is that it is trying to have strong portability guarantees when possible, and the current mapping mechanism prevents data races between the CPU and the GPU that would work on one system but not on another.

Also the fact that `readBuffer` puts data in arbitrary `ArrayBuffer` means that we need an extra copy to happen because in general `ArrayBuffers` are not shared memory between the GPU and content processes. (except `ArrayBuffers` you get from `GPUBuffer.getMappedRange`).

**Comment by fintelia** [comment]
https://github.com/gfx-rs/wgpu-rs/issues/694 is similar to this. The conclusion there was that a [`read_buffer`](https://docs.rs/wgpu/0.9.0/wgpu/util/struct.DownloadBuffer.html#method.read_buffer) convenience function could be added on top of the existing API.

**Comment by kvark** [comment]
Generally speaking, it doesn't have to be symmetric with `writeBuffer` because the target domain is not symmetric. CPU is always ahead of GPU, so passing data one way is easy, but the other way is hard. And applications need hundreds of little uploads per frame, while if there is a download at all, it's probably a single one. So any overhead for download (such as - if you just map a buffer yourself) is not as much important.

**Comment by fintelia** [comment]
> And applications need hundreds of little uploads per frame, while if there is a download at all, it's probably a single one. So any overhead for download (such as - if you just map a buffer yourself) is not as much important.

I disagree that applications will likely only do a single download per frame. Streaming results back from the GPU doesn't seem like that unreasonable of a thing to do. For instance, Terra generates heightmap tiles using a compute shader (and uses them on the GPU for rendering) but needs them back on the CPU to do collision detection.

**Comment by litherum** [comment]
+1 for readBuffer() as parallel to writeBuffer(). I argued for it a year ago when we added writeBuffer(), and I still think it's crazy we have one but not the other.

@Kangz is right, though, that readBuffer() shouldn't block, and would instead have to return a promise or something. (If the desire is "read whatever happens to be in the buffer _right now_ without blocking" then that would be racey and error-prone. It would also sometimes not even be implementable if the buffer data is private to the device and not accessible by the host.)

> CPU is always ahead of GPU

I don't quite understand this. Surely it is possible to enqueue work which takes longer to execute than it does to enqueue?

**Comment by kainino0x** [comment]
We've talked about it several times before, but we should really just add a synchronous map or synchronous readBuffer that's only present on workers.

IIUC, blocking APIs are OK on workers (like synchronous XHR), just not on the main thread. The thing that's not OK on workers is blocking forever (e.g. synchronous render/event loop architectures).

**Comment by kainino0x** [comment]
> > CPU is always ahead of GPU
> 
> I don't quite understand this. Surely it is possible to enqueue work which takes longer to execute than it does to enqueue?

This is simply saying that GPU work always happens after it's enqueued (not before).

Expressed another way, the CPU can enqueue general GPU work, but the GPU cannot enqueue general CPU work.

**Comment by benvanik** [comment]
TLDR: a synchronous try-map - even if restricted to workers - would make things dramatically better and may negate the immediate need for a readBuffer - but a readBuffer would be easier to use and require less poorly-utilized memory sitting around. If I had to pick I'd definitely vote a try-map because then readBuffer is easy (even if not efficient) to polyfill with copyBufferToBuffer, onSubmittedWorkDone, and try-map whereas the other way around emulating try-map is prohibitively expensive.

---

RE mapping in steady-state readback vs copies: unfortunately the approach doesn't scale alongside WebGPU's restrictions on exclusive whole-buffer access and async-only mapping. In a large dynamic system it's just not practical to have pools of mostly unused dedicated buffers hanging around - and gets more complicated when several libraries may be involved (each needing its own dedicated pools as the access lifetime management is so critical). I mean it's fine on workstations where you have near infinite memory or vulkan-like APIs where you can alias memory, but especially in the mobile web context we don't want dozens of MB of mostly unused buffers hanging around :) If there was subrange-aware mapping exclusivity checks or independent memory + buffer views that would avoid the need for dedicated buffers but also make the tracking the implementations have to do much more complex. For these scenarios where it's really "I have memory in this buffer on the GPU and I want that transferred efficiently to memory in this other buffer on the CPU" we ideally shouldn't require any extra buffers in user mode - even if on some implementations buffers are needed internally. Unless...

... we had blocking maps on workers! Big +1 to that! Because the atomics are only available there and we do fine-grained cross-thread synchronization we are going to be running everything off the main browser thread anyway and for people working with similar code operating on workers only makes life easier (a few fat calls to go main<->worker, vs these latency-sensitive high-frequency async mapping operations). Once you have blocking maps then the above approach gets more plausible as an application can use fewer large slabs of memory and subrange reads into them, however not knowing if a map would block or not makes it something useful only in exceptional cases (you don't want to be in the middle of a frame and not be able to tell if a call is going to block until the entire previous frame flushes, etc). A `GPUBuffer.tryMapSync` method, however, would be usable and could probably happen right on the main thread too: then a small pool of in-flight readback buffers and periodic attempts to map one would allow success to indicate completion. Still more wasteful memory-wise than if a single readback ringbuffer could have portions in-flight and portions being read but better than needing to read the tea leaves to guess pipeline depth.

Basic GPU->CPU pipelining with a `tryMapSync`:
```
// expensive CPU logic to issue work to the device; allowing the device to catch up a bit
resultBuffers = issueWork();
// before acquiring readback buffers try to reclaim a few; could do backpressure and bail if max reached
processReadbackBuffers();
// get a gpu buffer we can read into
tempBuffer = acquireReadbackBuffer(sz);
// enqueue copies of the result buffers into the readback buffer
enqueueCopyBuffer(resultBuffers[0], ..., tempBuffer, offset); offset += sizes[0];
enqueueCopyBuffer(resultBuffers[1], ..., tempBuffer, offset); offset += sizes[1];
...
queue.onSubmittedWorkDone().then(() => {
  // if the main loop driving this quiesced then this will catch the last cleanup work
  // importantly we have only one of these callbacks and can combine other work - vs one per copy/etc
  // it's also entirely possible this no-ops every time and that's fine (though a semaphore/fence would be way better :)
  processReadbackBuffers();
});
...
function processReadbackBuffers() {
  for (outstandingBuffer : reverse(outstandingBuffers)) {
    mapped = outstandingBuffer.tryMapSync(READ, ...);
    if (mapped) {
      // we now know all outstandingBuffers before this one are ready and mappable, can process them all
      // this assumes that implementations ensure tryMapSync will be cheap on those other buffers
    }
  }
}
```

Ideally a readBuffer wouldn't need to return a promise but instead onSubmittedWorkDone could be used to guard it; the promise returned from a readBuffer would just be used to emulate the kind of polling above in all but the simplest cases - at which point it's doing a non-trivial amount of work relative to what it should be (comparing an int/bool semaphore/fence): something in the driver getting called, going to the gpu process, hopping to site isolate, queuing work on the task list, at some highly variable point into the future calling into js, which then may call into wasm, to set a single bit, and then unwind. If you do 10, 20, 30, ... readbacks this all happens that many times. This would work but it's embarrassingly inefficient when the usage pattern is so well-known and straightforward :)

It gets worse in latency-sensitive situations where the CPU and GPU are not too far out of sync and even in native code it's quite important to avoid syscalls that may inadvertently yield your quantum as that ensures you introduce 10us-1ms+ of variance into your execution. Same way you don't want a setImmediate/promise in the middle of your game loop. Again this is where things like semaphores/fences can help as you can coalesce updates and avoid the additional work: if by the time your application code goes to run again and checks the value if you find 4 reads have complete in-between you can handle them together (or elide the 3 and take the last, etc) - whereas if you can't see forward into the future you have to take each high-variance callback in order. If there's a way to poll the promises for completion that'd be better as this coalescing could be approximated, but when thinking about low latency or low power systems you really don't want to be designing things that prevent your apps or the implementation from coalescing (or that force them to do their own coalescing poorly with imperfect information).

Given the implicit order on the WebGPU queue having readBuffer promises be neutered when they aren't used may be ok - then if you issued 10 readBuffers each returning a promise but only used the last one it may be efficient enough. My worry from a previous lifetime is that nothing is going to be looking to see if they aren't used and you'll get all the junk work happening except for the last-mile - but hopefully things have changed since I last dove into the browser internals :) Having an API that avoids the need for the entire browser stack across all browsers to be flawless in order to get relatively predictable performance seems better.

Comparing this to the classic double buffering trick in render loops: if I'm building frame T while frame T-1 is in-flight and I know by construction that frame T-2 has completed I'd like to be able to access frame T-2's resources without having to wait until frame T+1 because I did an async map (or dozens of async maps) and have to wait for them to all resolve. A synchronous try-map would allow for frame-wise polling and eager processing and the existing onSubmittedWorkDone callback would provide a good place for cleanup.

All that said, I think a native and efficient `GPUBuffer.tryMapSync` that allowed for failable mappings would make both pipelined host->device and device->host transfers respectively easier/more-efficient and possible. Though it'd be possible with a promise-based readBuffer to implement something like tryMapSync it won't be practically efficient and may still introduce copies/other issues (as mentioned with the ArrayBuffer target of a hypothetical readBuffer possibly being something the GPU cannot directly write into).

(sorry for the ramble, hope the background/scenario examples were useful :)

**Comment by kvark** [comment]
Thank you for the extensive analysis! It's hard to follow for me. Let me recap what I understood, and you can correct me (also, discussing this on `#WebGPU` matrix might be useful).

You are saying that returning a promise per `readBuffer` isn't good, because promises are heavy, and you expect to need this multiple times per frame. Therefore, you are suggesting to link the availability signal to `onSubmittedWorkDone`. And then we'd need some kind of a synchronous map, which would have to succeed if `onSubmittedWorkDone` was called (would it have to fail otherwise?). The most relevant discussion from the past that I could find is #594, although there might be more.

Immediate question here: with the current API, can't you schedule your reads into neighbouring parts of a single buffer, and just issue one `readBuffer` per frame?

**Comment by jdashg** [comment]
There were wars fought over this and I would like to defer revisiting those until after we have our MVP.

**Comment by Kangz** [comment]
The problem with a `tryMapSync` is that there's only intent to do a GPU -> CPU transfer at the time you ask "can I map now". However in this case the CPU and the GPU are in different processes (renderer and GPU process respectively). This causes two issues.

First it will be amazingly difficult to make the renderer process know at any one time whether a buffer is in use by the GPU or available for mapping. The renderer process would need to track when it gets submitted, but also what happens when there are errors, etc. It would essentially duplicated all validation in the renderer process.

Second there is work that must happen on the GPU process when a readback is initiated, for example the copy from the readback buffer to the shared memory between the two processes. It's important that the application signals its intent to readback so we can perform the copies early. Unless we eagerly copy to shmem on a regular basis, but that would have a big continuous cost as well.

Both of these issues can be fixed with a synchronous IPC between the renderer process and GPU process, but that would have a large fixed cost and unpredictability with the OS scheduling etc. I don't think you want it. This means that in all cases there is 1) some intent by the application that they want to readback so we can control the cost of copying to shmem 2) some asynchronous mechanism to say the readback is ready (this can be just "the GPU has finished execution past this point" like in WebGL with fences, and WebGPU's promises are well ordered so you get the same guarantees). This is basically what `mapAsync` is, and a similar `readBuffer` would just introduce additional copies.

I believe what would solve your problem better was if there was some way to run the browser event loop without yielding to it so that promises get fired (instead of inventing a try-blocking version of WebGPU promises-based API)

**Comment by kainino0x** [comment]
In discussing #4383 I suggested that it would be convenient to be able do something like readQuerySet for query set (timestamp results in particular).
